lib/uuid/uuid.c

```diff
@@ -0,0 +1,231 @@
+#define UUID_STR_POSITION_SECOND_HYPHEN (13U)
+#define UUID_STR_POSITION_THIRD_HYPHEN  (18U)
+#define UUID_STR_POSITION_FOURTH_HYPHEN (23U)
+
+#define UUID_POSITION_VERSION (6U)
+#define UUID_OFFSET_VERSION   (4U)
+#define UUID_MASK_VERSION     (0xF0U)
+#define UUID_POSITION_VARIANT (8U)
+#define UUID_OFFSET_VARIANT   (6U)
+#define UUID_MASK_VARIANT     (0xC0)
+
+#define UUID_V4_VERSION (4U)
+#define UUID_V4_VARIANT (2U)
+#define UUID_V5_VERSION (5U)
+#define UUID_V5_VARIANT (2U)
+
+#if defined(CONFIG_UUID_V4) || defined(CONFIG_UUID_V5)
+static void overwrite_uuid_version_and_variant(uuid_t uuid, uint8_t version, uint8_t variant)
+{
+	/* Clear the 'ver' and 'var' fields */
+	uuid[UUID_POSITION_VERSION] &= ~UUID_MASK_VERSION;
+	uuid[UUID_POSITION_VARIANT] &= ~UUID_MASK_VARIANT;
+	/* Update the 'ver' and 'var' fields */
+	uuid[UUID_POSITION_VERSION] |= (uint8_t)(version << UUID_OFFSET_VERSION);
+	uuid[UUID_POSITION_VARIANT] |= (uint8_t)(variant << UUID_OFFSET_VARIANT);
+}
+#endif
+
+static bool should_be_hyphen(uint8_t position)
+{
+	switch (position) {
+	case UUID_STR_POSITION_FIRST_HYPHEN:
+	case UUID_STR_POSITION_SECOND_HYPHEN:
+	case UUID_STR_POSITION_THIRD_HYPHEN:
+	case UUID_STR_POSITION_FOURTH_HYPHEN:
+		return true;
+	default:
+		return false;
+	}
+}
+
+#if defined(CONFIG_UUID_V4)
+void uuid_generate_v4(uuid_t out)
+{
+	/* Fill the whole UUID struct with a random number */
+	sys_rand_get(out, UUID_SIZE);
+	/* Update version and variant */
+	overwrite_uuid_version_and_variant(out, UUID_V4_VERSION, UUID_V4_VARIANT);
+}
+#endif
+
+#if defined(CONFIG_UUID_V5)
+int uuid_generate_v5(const uuid_t namespace, const void *data, size_t data_size, uuid_t out)
+{
+	int mbedtls_err = 0;
+	mbedtls_md_context_t ctx = {0};
+	const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
+	const size_t sha_1_bytes = 20;
+	uint8_t sha_result[sha_1_bytes];
+
+	mbedtls_md_init(&ctx);
+	mbedtls_err = mbedtls_md_setup(&ctx, md_info, 0);
+	/* Might return: MBEDTLS_ERR_MD_BAD_INPUT_DATA or MBEDTLS_ERR_MD_ALLOC_FAILED */
+	switch (mbedtls_err) {
+	case 0:
+		break;
+	case MBEDTLS_ERR_MD_BAD_INPUT_DATA:
+		return -EINVAL;
+	case MBEDTLS_ERR_MD_ALLOC_FAILED:
+		return -ENOMEM;
+	default:
+		return -ENOTSUP;
+	}
+	mbedtls_err = mbedtls_md_starts(&ctx);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_update(&ctx, namespace, UUID_SIZE);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_update(&ctx, data, data_size);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_finish(&ctx, sha_result);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_md_free(&ctx);
+
+	/* Store the computed SHA1 in the out struct */
+	for (size_t i = 0; i < UUID_SIZE; i++) {
```

Comment:
```suggestion
	for (int i = 0; i < UUID_SIZE; i++) {
```

Minor; `UUID_SIZE` is currently defined as an `int`, and we should consider using the same type here. 

`UUID_SIZE` and the other macros can be defined to be `unsigned int` by appending `U` to the value, which would make this
```suggestion
	for (unsigned int i = 0; i < UUID_SIZE; i++) {
```
instead